doc/debugger.rst

```diff
@@ -1773,6 +1811,161 @@ target endianness and interpret the bytecode based on that.
+Each value pushed will be sent as a dvalue in the message.  So if you push two
+strings, "foo" and "bar", the client will see ``NFY 7 "foo" "bar" EOM``.
+
+AppRequest is used to make requests to the target which are not directly
+related to Ecmascript execution and may be implementation-dependant.  For
+example, an AppRequest might be used to:
+
+* Change the frame rate of a game engine
+
+* Reset/reboot an embedded target device while debugging
+
+* Perform or trigger software or script updates
+
+A target that wishes to support AppRequest should attach the debugger using the
+``duk_debugger_attach_custom()`` API and provide a request callback.  When an
+AppRequest is received, the request callback is invoked with the contents of
+the message on the value stack, and may push its own values to be sent in
+reply.
+
+This is a do-nothing request callback::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+        /* nop */
+        return 0;
+    }
+
+The above callback simply responds with ``REP EOM`` (an empty reply) to all
+requests.  This is admittedly not very useful and a real implementation should
+process the values it receives on the stack, push its own values to send in
+reply, and return a non-negative integer indicating how many values it pushed.
+
+Here is a slightly more useful implementation::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+        /* Must use negative stack indices to access dvalues */
+        const char* cmdname = duk_get_string(ctx, -nvalues + 0);
+        
+        if (strcmp(cmdname, "VersionInfo") == 0) {
+            /* Return a positive integer to send a REP containing values pushed
+             * to the stack. The return value indicates how many dvalues you
+             * are including in the response.
+             */
+            
+            duk_push_string(ctx, "My Awesome Program");
+            duk_push_int(ctx, 81200);  /* ver. 8.12.0 */
+            
+            return 2;  /* 2 dvalues */
+        } else {
+            /* Return -1 to send an ERR reply. The value on top of the stack
+             * will be string coerced and used for the error text.
+             */
+            
+            duk_push_string(ctx, "unrecognized AppRequest command name");
+            return -1;
+        }
+    }
+
+If no request callback is provided at attach, AppRequest will be treated as an
+unsupported command, eliciting an ERR reply from Duktape saying so.  Such a
+target is still free to send AppNotify messages, however.
+
+As a precaution, the target should try to avoid sending structured values such
+as JS objects in notify messages as their heap pointers may become stale by the
+time the client receives them.  This is especially true for notifications sent
+while the target is running.  It's better to stick to primitives which have
+unique dvalue representations, i.e. numbers, strings and buffers.  If a
```

Comment:
The complete list would actually be: `undefined`, `null`, boolean, number, string, buffer, pointer.

If that's too much then maybe change `i.e.` to `e.g.`, I'd still add the boolean though.